### ISSUE TITLE
Training and inference details

### ISSUE THREAD

**Issue by DarrenQu** [yifanlu0227/HEAL#22]
Thank you for your outstanding work. About the data presented in Table 2 of your paper, could you please provide more training and inference details? Like, how did you train the models to achieve the results shown in the table? Are there any pre-trained models available that I can use?
![image](https://github.com/user-attachments/assets/94785013-3af4-42c8-8223-5a3f1d61932b)

**Comment by Chinese-Coding** [comment]
# Chinese
我是按照作者提供的 README 进行训练的，因为我使用了 spconv2.x 的版本，所以为并没有使用作者提供的 checkpoint。

我的使用到的训练命令和测试结果如下所示。我很希望您也能将您的训练命令和测试结果分享一下，方便交流。

# English
I trained according to the README provided by the author. Since I used the spconv2.x version, I did not use the checkpoint provided by the author.

The training commands and test results I used are shown below. I hope you can also share your training commands and test results for easy communication.

# Result (训练结果)
## m1 based
### stag1
#### train
```bash
CUDA_VISIBLE_DEVICES=1 python opencood/tools/train.py -y None --model_dir opencood/logs/origin_heal/m1_based/stage1
```
#### test and result
* early
```bash
CUDA_VISIBLE_DEVICES=1 python opencood/tools/inference.py --model_dir opencood/logs/origin_heal/m1_based/stage1 --fusion_method early
The Average Precision at IOU 0.3 is 0.96, The Average Precision at IOU 0.5 is 0.96, The Average Precision at IOU 0.7 is 0.93
```

* intermediate
```bash
CUDA_VISIBLE_DEVICES=1 python opencood/tools/inference.py --model_dir opencood/logs/origin_heal/m1_based/stage1 --fusion_method intermediate
The Average Precision at IOU 0.3 is 0.96, The Average Precision at IOU 0.5 is 0.96, The Average Precision at IOU 0.7 is 0.93
```

* late
```bash
CUDA_VISIBLE_DEVICES=1 python opencood/tools/inference.py --model_dir opencood/logs/origin_heal/m1_based/stage1 --fusion_method late
The Average Precision at IOU 0.3 is 0.96, The Average Precision at IOU 0.5 is 0.96, The Average Precision at IOU 0.7 is 0.92
```

* no
```bash
CUDA_VISIBLE_DEVICES=1 python opencood/tools/inference.py --model_dir opencood/logs/origin_heal/m1_based/stage1 --fusion_method no
AttributeError: 'IntermediateheterFusionDataset' object has no attribute 'post_process_no_fusion_uncertainty'
```

* no_w_uncertainty
```bash
CUDA_VISIBLE_DEVICES=1 python opencood/tools/inference.py --model_dir opencood/logs/origin_heal/m1_based/stage1 --fusion_method no_w_uncertainty
AttributeError: 'IntermediateheterFusionDataset' object has no attribute 'post_process_no_fusion_uncertainty'
```

### stag2
#### train
#### test and result

**Comment by yifanlu0227** [comment]
<img width="1030" alt="Screenshot 2024-07-16 at 10 27 39" src="https://github.com/user-attachments/assets/c73e1010-5639-477d-aa1b-2dc25e019cd7">

## Checkpoint
Hi! Firstly you can find our checkpoints in [HuggingfaceHub](https://huggingface.co/yifanlu/HEAL/tree/main). The `HEAL` models for OPV2V-H and DAIR-V2X are stored as `HEAL_OPV2V.zip` and `HEAL_DAIR.zip`, which you can used to infer directly. Just notice they are trained with spconv 1.2.1, not compatible with spconv 2.x. 

## Inference
To reproduce the result in the table, you will use `opencood/tools/inference_heter_in_order.py`. It adopts a sequential assignment json file (e.g. `modality_assign/opv2v_4modality_in_order.json`) to make sure agent types are integrated in order; and control the agent types (as well as the agent number) in the scene via `--use_cav`. 

`--use_cav` accepts a list. `--use_cav [2,3,4]` means we will set the agent type number to 2 / 3 / 4 in the scene to do the inferences, which correspond to the experiment table (based on $L_P^{(64)}$, gradually adding $C_E^{(384)}$, $L_S^{(32)}$, $C_R^{(336)}$).

`python opencood/tools/inference_heter_in_order.py --model_dir opencood/logs/HEAL_OPV2V/final_infer --use_cav [2,3,4]`

## Training
Follow the instruction in [README](https://github.com/yifanlu0227/HEAL#heals-train-command).

**Comment by yifanlu0227** [comment]
Since @Chinese-Coding find some bugs in the model configuration for `SECOND` #20 , it is not recommended that you develop new models based on existing checkpoints (spconv 1.2.1). 

You would enjoy much more convenient installation with `spconv 2.x` and training from scratch is not that slow.

**Comment by scz023** [comment]
> <img alt="Screenshot 2024-07-16 at 10 27 39" width="1030" src="https://private-user-images.githubusercontent.com/45688237/348917134-c73e1010-5639-477d-aa1b-2dc25e019cd7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI1ODU0OTIsIm5iZiI6MTcyMjU4NTE5MiwicGF0aCI6Ii80NTY4ODIzNy8zNDg5MTcxMzQtYzczZTEwMTAtNTYzOS00NzdkLWFhMWItMmRjMjVlMDE5Y2Q3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODAyVDA3NTMxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU0NWY2NDdjZGQwNzQ5YzEwYzkwOThmZDdjMGQ2NGExYjIyYTIzYjI3NzFlYzljZWM5ODViODE0ZDFiMmNmOTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.2xj_s-kRMYOiqPXWHCMabWWQ562geZye2SdFoIgbu8k">
> ## Checkpoint
> Hi! Firstly you can find our checkpoints in [HuggingfaceHub](https://huggingface.co/yifanlu/HEAL/tree/main). The `HEAL` models for OPV2V-H and DAIR-V2X are stored as `HEAL_OPV2V.zip` and `HEAL_DAIR.zip`, which you can used to infer directly. Just notice they are trained with spconv 1.2.1, not compatible with spconv 2.x.
> 
> ## Inference
> To reproduce the result in the table, you will use `opencood/tools/inference_heter_in_order.py`. It adopts a sequential assignment json file (e.g. `modality_assign/opv2v_4modality_in_order.json`) to make sure agent types are integrated in order; and control the agent types (as well as the agent number) in the scene via `--use_cav`.
> 
> `--use_cav` accepts a list. `--use_cav [2,3,4]` means we will set the agent type number to 2 / 3 / 4 in the scene to do the inferences, which correspond to the experiment table (based on LP(64), gradually adding CE(384), LS(32), CR(336)).
> 
> `python opencood/tools/inference_heter_in_order.py --model_dir opencood/logs/HEAL_OPV2V/final_infer --use_cav [2,3,4]`
> 
> ## Training
> Follow the instruction in [README](https://github.com/yifanlu0227/HEAL#heals-train-command).

Hello, I have another question, is Table2 obtained by using `opv2v_4modality_in_order.json`?
And, is Table3 configured using `opv2v_4modality.json`? So ego can be any mode (unlike Table2)?
![c6a43dbbaecd9fc25659d727111012f](https://github.com/user-attachments/assets/493113b4-cdb1-427f-baa5-3e57847f97a9)

**Comment by scz023** [comment]
> Hi @scz023 The table 3 is also using `opv2v_4modality_in_order.json`, as stated in the paper
> 
> > Table 3: Heterogeneous type agents are added in the order presented from left to right in each type combination.
> 
> You can get the metrics by
> 
> ```
> python opencood/tools/inference_heter_in_order.py --model_dir opencood/logs/HEAL_m1_based/final_infer  --use_cav [4] --lidar_degrade
> ```

Thank you!
Beseides, how are the baseline results obtained?
Is it to use `opencood/hypes_yaml/opv2v/MoreModality/x_modality_end2end_training/xxx.yaml` directly for multi-modal end-to-end training, and if so, how are the baseline results in Table1 obtained?

**Comment by yifanlu0227** [comment]
> Is it to use opencood/hypes_yaml/opv2v/MoreModality/x_modality_end2end_training/xxx.yaml directly for multi-modal end-to-end training

Yes! You also infer with `python opencood/tools/inference_heter_in_order.py --model_dir opencood/logs/<your baseline experiment log>`. 

Note that the `--use_cav` argument should align with the modality & model count in the input YAML file. Baselines need to train the new model from scratch when new modality & model is added.

For example, 

1. You train `opencood/hypes_yaml/opv2v/MoreModality/3_modality_end2end_training/m1m2m3_attfuse.yaml`, you are supposed to use `--use_cav [3]`. 
2. You train `opencood/hypes_yaml/opv2v/MoreModality/4_modality_end2end_training/m1m2m3m4_attfuse.yaml`, you are supposed to use `--use_cav [4]`.

**Comment by scz023** [comment]
> Note that the `--use_cav` argument should align with the modality & model count in the input YAML file. Baselines need to train the new model from scratch when new modality & model is added.

Got it, thank you！

**Comment by gpanaretou** [comment]
> Hi [@scz023](https://github.com/scz023) The table 3 is also using `opv2v_4modality_in_order.json`, as stated in the paper
> 
> > Table 3: Heterogeneous type agents are added in the order presented from left to right in each type combination.
> 
> You can get the metrics by
> 
> ```
> python opencood/tools/inference_heter_in_order.py --model_dir opencood/logs/HEAL_m1_based/final_infer  --use_cav [4] --lidar_degrade
> ```

Hello, about this, I can't recreate the results. While the individual agents' training and collaboration base match your results, for the final heterogeneous result, I am getting different results when using cav's [2, 3, 4]. I have identified that the issue starts with m2 (camera agent), but I am unsure what is happening, as I have also used the configs you provide in HuggingFace.

Have you encountered a possible issue with this? Could it be a dataset misconfiguration?

These are the results I am getting for the opv2v dataset with --use_cav [2]
```
ap30: 0.6798738780324204
ap_50: 0.6679914326250905
ap_70: 0.5478710853206901
mpre_50:
...
```

**Comment by gpanaretou** [comment]
Yes, I have isolated the problem. Something seems to be the issue with the M2 agent, and most likely the EfficientNet (not 100% on this). 

I have replaced my testing of 2 agents, instead of m1 + m2, with results:
```
ap30: 0.6798738780324204
ap_50: 0.6679914326250905
ap_70: 0.5478710853206901
```

To m1 + m4 with the following results:
```
The Average Precision at IOU 0.3 is 0.84, The Average Precision at IOU 0.5 is 0.82, The Average Precision at IOU 0.7 is 0.70
```
